src/openqaoa-core/qaoa_components/ansatz_constructor/baseparams.py

```diff
@@ -295,13 +269,164 @@ def __repr__(self):
+                input_object, block_type
+            )
+        elif isinstance(input_object, list):
+            for each_gate in input_object:
+                if isinstance(each_gate, RotationGateMap):
+                    each_gate.gate_label.update_gatelabel(
+                        new_gatemap_type=block_type
+                    )
+                else:
+                    raise TypeError(
+                        f"Input gate is of unsupported type {type(each_gate)}."
+                        "Only RotationGateMaps are supported"
+                    )
+            block = input_object
+        else:
+            raise ValueError(
+                "The input object defining mixer should be a List of RotationGateMaps or type Hamiltonian"
+            )
+        return block
+
+    def reorder_gates_block(self, gates_block, layer_number):
+        """Update the qubits that the gates are acting on after application
+        of SWAPs in the cost layer
+        """
+        for gate in gates_block:
+            
+            if layer_number%2 == 0:
+                mapping = self.final_mapping         
+                gate.qubit_1 = mapping[gate.qubit_1]
+                if gate.gate_label.n_qubits == 2:
+                    gate.qubit_2 = mapping[gate.qubit_2]
+            else:
+                pass
+            
+        return gates_block
+                      
+    @staticmethod
+    def route_gates_list(gates_to_route: List[GateMap], device:DeviceBase, routing_function: Callable) -> List[GateMap]:
+        """
+        Apply qubit routing to the abstract circuit gate list
+        based on device information
+        
+        Parameters
+        ----------
+        gates_to_route: `List[GateMap]`
+            The gates to route
+        device: `DeviceBase`
+            The device on which to run the circuit
+        routing_function: `Callable`
+            The function that accepts as input the device, problem, initial_mapping and
+            outputs the list of gates with swaps
+        """
+        problem_to_solve = [[gate.qubit_1, gate.qubit_2] for gate in gates_to_route
+                            if gate.gate_label.n_qubits == 2]
+        (
+            gate_list_indices,
+            swap_mask,
+            initial_physical_to_logical_mapping,
+            final_mapping
+        ) = routing_function(device, problem_to_solve)
+        
+        print("Initial mapping", initial_physical_to_logical_mapping)
+        print("Final mapping", final_mapping)
+        
+        # for gate in gates_to_route:
+        #     gate.qubit_1 = initial_physical_to_logical_mapping[gate.qubit_1]
+        #     if gate.gate_label.n_qubits == 2:
+        #         gate.qubit_2 = initial_physical_to_logical_mapping[gate.qubit_2]
+            
+        original_qubits_to_gate_mapping = {
+            (gate.qubit_1, gate.qubit_2): gate for gate in gates_to_route
+            if gate.gate_label.n_qubits == 2
+        }
```

Comment:
I think its possible to use `original_qubits_to_gate_mapping ` to get `problem_to_solve` via `[list(each_pair) for each_pair in original_qubits_to_gate_mapping.keys()]`

Reply:
already fixed, thanks!